### PR TITLE
Fixed ScriptsToProcess when -Version param is used in Import-Module

### PR DESCRIPTION
**Pull request by anmenaga**
This is fix for #3738 "ScriptsToProcess not honored when Import-Module is passed -Version param".

The bug was cased by existing check in LoadModule functions:

> If  MinimumVersion/RequiredVersion/MaximumVersion has been specified, then only try to process manifest modules...
> If the -Version flag was specified, don't look for non-manifest modules.

When loading usual module files, this check is valid;
However, same LoadModule function is used to load ScriptsToProcess - and this when this check was 
triggered and ScriptsToProcess were not executed.

The fix is to work around that condition for ScriptsToProcess in the same way it is currently done for NestedModules - temporary reset the version info for the duration of processing ScriptsToProcess / NestedModules.

Test results before the fix:
befoerfix

Test results after the fix:
afterfix

### CODE REVIEW

File: src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs [hunk 1]
```diff
@@ -2768,38 +2768,58 @@ internal bool LoadModuleManifestData(ExternalScriptInfo scriptInfo, ManifestProc
...
-                        foreach (string detectedFunction in module.ExportedFunctions.Keys)
+                        // If we're in analysis, add the detected exports to this module's
+                        // exports
+                        if (found && (ss == null))
                         {
-                            manifestInfo.AddDetectedFunctionExport(detectedFunction);
-                        }
+                            foreach (string detectedCmdlet in module.ExportedCmdlets.Keys)
+                            {
+                                manifestInfo.AddDetectedCmdletExport(detectedCmdlet);
```
Comment (adityapatwardhan) [inline src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs]: AddDetected* methods from PSModuleInfo in most cases seem to be called multiple times using a for loop. Will it make sense to add an overload which accepts IEnumerable?

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1 [hunk 2]
```diff
@@ -22,3 +22,52 @@
...
+    It "Verify ScriptsToProcess are executed for top-level module when -Version is specified" {
+        Import-Module .\module1.psd1 -Version 1.0
+        Get-Content out.txt | Should Be '1'
+    }
+
+    It "Verify ScriptsToProcess are executed for top-level and nested module when -Version is specified" {
+        Import-Module .\module2.psd1 -Version 1.0
+        Get-Content out.txt | Should Be '21'
+    }
+}
```
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: Add new line at the end.

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1 [hunk 3]
```diff
@@ -22,3 +22,52 @@
...
+
+        "1 | Out-File out.txt -Append -NoNewline" | Out-File script1.ps1
+        "2 | Out-File out.txt -Append -NoNewline" | Out-File script2.ps1
+        New-ModuleManifest module1.psd1 -ScriptsToProcess script1.ps1
+        New-ModuleManifest module2.psd1 -ScriptsToProcess script2.ps1 -NestedModules module1.psd1
+    }
+
+    AfterAll {
+        Pop-Location
+        #Remove-Item $moduleRootPath -Recurse -Force -ErrorAction SilentlyContinue
```
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: Remove commented line.

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1 [hunk 4]
```diff
@@ -22,3 +22,52 @@
...
+        New-Item out.txt -ItemType File -Force | Out-Null
+    }
+
+    AfterEach {
+        $m = @('module1','module2','script1','script2')
+        remove-module $m -Force -ErrorAction SilentlyContinue
+        Remove-Item out.txt -Force -ErrorAction SilentlyContinue
+    }
+    
+    It "Verify ScriptsToProcess are executed for top-level module" {
```
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: You may want to take a look at TestCases parameter of Pester It.

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1 [hunk 5]
```diff
@@ -22,3 +22,52 @@
         (Get-Module -Name $moduleName).Name | Should Be $moduleName
     }
 }
+
+Describe "Import-Module with ScriptsToProcess" -Tags "CI" {
+    
+    BeforeAll {
+        $moduleRootPath = Join-Path $TestDrive 'TestModules'
+        New-Item $moduleRootPath -ItemType Directory -Force | Out-Null
+        Push-Location $moduleRootPath
```
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: Why do you need to push location? Can you use full paths for all cmdlets like New-Item, New-ModuleManifest etc?

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1 (hunk 3 above)
Comment (anmenaga) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: Fixed.

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1 (hunk 2 above)
Comment (anmenaga) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: Fixed.

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1 (hunk 4 above)
Comment (anmenaga) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: Thanks for suggestion; I did give it a try. I believe that semantic difference between these tests is larger than what TestCases parameter is aimed for; plus I don't like that TestCases parameter has worse logging (same test name repeated several times vs individual, descriptive test names in case of individual 'It's).

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1 (hunk 5 above)
Comment (anmenaga) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: I chose push location specifically because I didn't want to deal with full paths. :)

File: src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs [hunk 6]
```diff
@@ -2768,38 +2768,58 @@ internal bool LoadModuleManifestData(ExternalScriptInfo scriptInfo, ManifestProc
...
+                    BaseGuid = null;
+
+                    foreach (string scriptFile in scriptsToProcess)
                     {
-                        foreach (string detectedCmdlet in module.ExportedCmdlets.Keys)
-                        {
-                            manifestInfo.AddDetectedCmdletExport(detectedCmdlet);
-                        }
+                        bool found = false;
+                        PSModuleInfo module = LoadModule(scriptFile,
```
Comment (anmenaga) [inline src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs]: That is actually existing code; I didn't change it; looks like Github's diff'er has room for improvement :)  - see picture below.
Anyway, there are multiple LoadModule calls like this one throughout the file, so if refactoring is to be done - it has to be done to all of them - and that is out of scope of this bugfix.

diff

File: src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs (hunk 1 above)
Comment (anmenaga) [inline src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs]: Maybe. But as previous comment - this would be refactoring of unrelated code - out of scope of this bug fix.

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1 (hunk 4 above)
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: You can have custom test case names for individual test cases. Have look at: https://github.com/PowerShell/PowerShell/blob/02b5f357a20e6dee9f8e60e3adb9025be3c94490/test/powershell/engine/ParameterBinding/NullableBooleanDCR.Tests.ps1

Here the parameters will be parameter splat for Import-Module, and expected result, test name.

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1 (hunk 4 above)
Comment (anmenaga) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: That is actually cool; thank you!
I've updated tests.

Comment (anmenaga) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: Good idea; I've updated tests.

**Review by adityapatwardhan — APPROVED** [review]
